packages/core/nano-server/src/middleware/cors-helper.ts

```diff
@@ -0,0 +1,30 @@
+import type {AlwatrConnection} from '../nano-server';
+import type {CrossOriginResourceSharingHeaders} from '../type';
+
+export function setCORSHelperHeader(connection: AlwatrConnection, option: CrossOriginResourceSharingHeaders): void {
+  const allowedOrigin = option.accessControlAllowOrigin;
+  const requestOrigin = connection.incomingMessage.headers.origin;
+
+  if (requestOrigin === undefined) return;
+
+  let allowed = false;
+
+  if (allowedOrigin === '*') {
+    connection.serverResponse.setHeader('Access-Control-Allow-Origin', allowedOrigin);
+    allowed = true;
+  } else if (Array.isArray(allowedOrigin)) {
+    if (allowedOrigin.includes(requestOrigin)) {
+      connection.serverResponse.setHeader('Access-Control-Allow-Origin', requestOrigin);
+      if (allowedOrigin.length > 1) {
+        connection.serverResponse.setHeader('Vary', 'Origin');
```

Comment:
> If the server specifies a single origin (that may dynamically change based on the requesting origin as part of an allowlist) rather than the "*" wildcard, then the server should also include Origin in the [Vary](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary) response header to indicate to clients that server responses will differ based on the value of the [Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) request header.

https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#access-control-allow-origin